tests/Form/DataTransformerResolverTest.php

```diff
@@ -0,0 +1,221 @@
+        $this->fieldDescription->getType()->willReturn('choice');
+        $this->fieldDescription->getOption('class')->willReturn(\stdClass::class);
+        $this->fieldDescription->getTargetModel()->willReturn(\DateTime::class);
+        $this->fieldDescription->getTargetEntity()->willReturn(\DateTime::class);
+
+        $this->assertNull($this->resolve());
+    }
+
+    public function testResolveChoice(): void
+    {
+        $newId = 1;
+        $className = \stdClass::class;
+        $object = new \stdClass();
+
+        $this->fieldDescription->getOption('data_transformer')->willReturn(null);
+        $this->fieldDescription->getType()->willReturn('choice');
+        $this->fieldDescription->getOption('class')->willReturn($className);
+        $this->fieldDescription->getTargetModel()->willReturn($className);
+
+        $this->modelManager->find($className, $newId)->willReturn($object);
+
+        $dataTransformer = $this->resolve();
+
+        $this->assertInstanceOf(ModelToIdTransformer::class, $dataTransformer);
+        $this->assertSame($object, $dataTransformer->reverseTransform($newId));
+    }
+
+    /**
+     * @dataProvider provideFieldTypes
+     */
+    public function testCustomGlobalTransformers(string $fieldType): void
+    {
+        $customDataTransformer = new CallbackTransformer(static function ($value): string {
+            return (string) (int) $value;
+        }, static function ($value): bool {
+            return filter_var($value, FILTER_VALIDATE_BOOLEAN);
+        });
+
+        $this->fieldDescription->getOption('data_transformer')->willReturn(null);
+        $this->fieldDescription->getType()->willReturn($fieldType);
+
+        $this->resolver = new DataTransformerResolver([
+            $fieldType => $customDataTransformer, // override predefined transformer
+        ]);
+
+        $dataTransformer = $this->resolve();
+
+        $this->assertInstanceOf(DataTransformerInterface::class, $dataTransformer);
+        $this->assertSame($customDataTransformer, $dataTransformer);
+    }
+
+    /**
+     * @dataProvider provideFieldTypes
+     */
+    public function testAddCustomGlobalTransformer(string $fieldType): void
+    {
+        $customDataTransformer = new CallbackTransformer(static function ($value): string {
+            return (string) (int) $value;
+        }, static function ($value): bool {
+            return filter_var($value, FILTER_VALIDATE_BOOLEAN);
+        });
+
+        $this->fieldDescription->getOption('data_transformer')->willReturn(null);
+        $this->fieldDescription->getType()->willReturn($fieldType);
+
+        $this->resolver->addCustomGlobalTransformer($fieldType, $customDataTransformer);
+
+        $dataTransformer = $this->resolve();
+
+        $this->assertInstanceOf(DataTransformerInterface::class, $dataTransformer);
+        $this->assertSame($customDataTransformer, $dataTransformer);
+    }
+
+    protected function resolve(): ?DataTransformerInterface
```

Comment:
Can't this method be private?

Reply:
Yes. Must be private. Now i don't remember why i did this.